sticht/metric.py

```diff
@@ -0,0 +1,102 @@
+# Copyright 2019-2022 Yelp Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import abc
+import threading
+from typing import Any
+from typing import Callable
+from typing import List
+from typing import Optional
+from typing import Tuple
+
+import pytimeparse
+
+from sticht.slack import SlackDeploymentProcess
+
+
+# TODO: Rethink if MetricDemultiplexer is required
+
+
+class MetricWatcher:
+    def __init__(
+        self,
+        metric: Any,
+        callback: Callable[['MetricWatcher'], Any],
+        start_timestamp: float,
+        label: str,
+        max_duration: float,
+    ) -> None:
+        self.metric = metric
+        self.window: List[Tuple[float, float]] = []
+        self.max_duration = max_duration
+        self.failing: Optional[bool] = None
+        self.bad_after_mark: Optional[bool] = None
+        self.bad_before_mark: Optional[bool] = None
+        self.callback = callback
+        self.start_timestamp = start_timestamp
+        self.label = label
```

Comment:
it doesn't look like mfd is actually directly creating SLOWatchers right now - maybe we just have the parent class for all these be as empty as possible?

e.g., i dunno if we're really going to be using all the window-based stuff for the new autorollbacks - if i understand our vision correctly, we're more worried about how many of our new rules are allowed to be failing at a given time + how many are failing by the end of a deploy (and not necessarily keeping around a moving window of datapoints and when they were good/bad)